eslint.config.js

```diff
@@ -16,19 +16,22 @@ const eslintPluginTestsRecommendedConfig = require("eslint-plugin-eslint-plugin/
 const globals = require("globals");
 const eslintConfigESLintCJS = require("eslint-config-eslint/cjs");
 const eslintConfigESLintFormatting = require("eslint-config-eslint/formatting");
+const json = require("@eslint/json");
 
 //-----------------------------------------------------------------------------
 // Helpers
 //-----------------------------------------------------------------------------
 
 const INTERNAL_FILES = {
-    CLI_ENGINE_PATTERN: "lib/cli-engine/**/*",
-    LINTER_PATTERN: "lib/linter/**/*",
-    RULE_TESTER_PATTERN: "lib/rule-tester/**/*",
-    RULES_PATTERN: "lib/rules/**/*",
-    SOURCE_CODE_PATTERN: "lib/source-code/**/*"
+    CLI_ENGINE_PATTERN: "lib/cli-engine/**/*.js",
+    LINTER_PATTERN: "lib/linter/**/*.js",
+    RULE_TESTER_PATTERN: "lib/rule-tester/**/*.js",
+    RULES_PATTERN: "lib/rules/**/*.js",
+    SOURCE_CODE_PATTERN: "lib/source-code/**/*.js"
```

Comment:
This would break the following logic in `createInternalFilesPatterns`:

https://github.com/eslint/eslint/blob/a7016a50d88011f279d52b9355a5662e561c414c/eslint.config.js#L55-L56

I think we could update the regex to `/\*\*\/\*\.js$/u` to fix this.

Reply:
How would I be able to tell if something broke?

Reply:
This is intended to allow accessing submodules only through their main exports (`index.js` files), enforced by `n/no-restricted-require` rule.

For example,`lib/api.js` is allowed to use `lib/linter/index.js`, but nothing else from `lib/linter/` directly.

`$ node bin/eslint --print-config lib/api.js` in the `main` branch:

```
"n/no-restricted-require": [
      2,
      [
        {
          "name": [
            "C:\\projects\\eslint\\lib\\cli-engine\\**\\*",
            "!C:\\projects\\eslint\\lib\\cli-engine\\index.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\linter\\**\\*",
            "!C:\\projects\\eslint\\lib\\linter\\index.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\rule-tester\\**\\*",
            "!C:\\projects\\eslint\\lib\\rule-tester\\index.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\rules\\**\\*",
            "!C:\\projects\\eslint\\lib\\rules\\index.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\source-code\\**\\*",
            "!C:\\projects\\eslint\\lib\\source-code\\index.js"
          ]
        }
      ]
    ]
```

Adding something like `const NodeEventGenerator = require("./linter/node-event-generator");` to `lib/api.js` produces a lint error from `n/no-restricted-require`.

`$ node bin/eslint --print-config lib/api.js` in this branch:

```
"n/no-restricted-require": [
      2,
      [
        {
          "name": [
            "C:\\projects\\eslint\\lib\\cli-engine\\**\\*.js",
            "!C:\\projects\\eslint\\lib\\cli-engine\\**\\*.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\linter\\**\\*.js",
            "!C:\\projects\\eslint\\lib\\linter\\**\\*.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\rule-tester\\**\\*.js",
            "!C:\\projects\\eslint\\lib\\rule-tester\\**\\*.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\rules\\**\\*.js",
            "!C:\\projects\\eslint\\lib\\rules\\**\\*.js"
          ]
        },
        {
          "name": [
            "C:\\projects\\eslint\\lib\\source-code\\**\\*.js",
            "!C:\\projects\\eslint\\lib\\source-code\\**\\*.js"
          ]
        }
      ]
    ]
  }
```

This configuration for `n/no-restricted-require` basically doesn't restrict anything, so adding `const NodeEventGenerator = require("./linter/node-event-generator");` to `lib/api.js` does not produce a lint error from `n/no-restricted-require`.

Reply:
Gotcha. I'll take a look.